doc/README.md

```diff
@@ -0,0 +1,421 @@
+    ┃ (0x02)  ┃ ┗━━━━━━━━━━━━━━━┻━━━━━━┻━━━━━━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━┛ ┃
+    ┗━━━━━━━━━┻━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+```
+
+When writing app data or apps to backup storage,
+the authenticated associated data (AAD) contains the backup version as the only byte
+(to prevent downgrade attacks) as it would otherwise not be authenticated.
+Other data is not included as renaming and swapping out files is made impossible
+by their file name being the content hash.
+
+## Content-defined chunking
+
+We use FastCDC ([PDF](https://www.usenix.org/system/files/conference/atc16/atc16-paper-xia.pdf))
+to split ZIP files and TAR streams we receive from the system into re-usable chunks.
+We aim for an average chunk size of 3 MiB,
+because in our tests it presented a good trade-off
+between deduplication ratio and number of small chunks.
+Data smaller than 1.5 MiB will not be chunked further and be left as a single chunk.
+
+TODO: Can we do something to prevent watermarking attacks against single chunk files,
+      like padding somewhat them after compression, but before encryption?
+
+The FastCDC algorithm uses a gear table containing 256 random integers with 31 bits.
+When this table changes, the resulting chunks will be different.
+Hence, every repository always uses the same gear table.
+However, to make watermarking attacks harder,
+we use the "app backup gear table key" that gets derived from our main key
+to deterministically compute a gear table using AES CTR to cipher 32 null bytes with a null IV.
+
+TODO: Is it safe to compute the gear table like this? We do derive a dedicated key for this
+      that we don't use for anything else.
+      Still, the IV as well as the encrypted bytes are known to the attacker.
+      Also, it may be be possible to reverse the gear table with chosen plaintext attacks.
+
+## Operations
+
+### Backup
+
+* download all snapshots to get a mapping of all plaintext hashes (chunk IDs) to storage IDs
+* optionally: get a list of all blobs and their storage IDs to confirm they (still) exist
+* if APK backup wasn't disabled, for each app:
+  * get version code and check if we already have an APK with that version code in a snapshot
+  * if APK was backed up already, re-use list of plaintext hashes for current snapshot
+  * if APK was not yet backed up, put it (and its splits) through the chunker
+  * chunks already in the repository are not uploaded again, only their hash recorded
+  * new chunks get compressed, encrypted and hashed to determine their storage ID, then uploaded 
+* for each app (the system allows us to back up):
+  * app data we receive as a tar stream from the system gets chunked
+    * chunks already in the repository are not uploaded again, only their hash recorded
+    * new chunks get compressed, encrypted and hashed to determine their storage ID, then uploaded
+  * remember ordered list of chunk IDs for the app (and its APKs)
+* add newly packed chunks to local index
```

Comment:
[Outdated?] Later (line 390 and beyond), it is written that pack files and indexes are not used. So what does this line and the lines below mean, or are they outdated?

Reply:
good catch. a left over from a time the design had still packs and indexes